src/plugins/expressions/common/ast/build_function.ts

```diff
@@ -0,0 +1,170 @@
+   *
+   * @param name The name of the argument to add.
+   * @param value The value of the argument. Can be a single value or an array.
+   * @return `this`
+   */
+  addArgument: <A extends FunctionArgName<Fn>>(name: A, value: FunctionArgs<Fn>[A]) => this;
+  /**
+   * Retrieves an existing argument by name.
+   * Useful when you want to retrieve the current array of args and add
+   * something to it before calling `replaceArgument`.
+   *
+   * @param name The name of the argument to retrieve.
+   * @return `ExpressionAstArgument[]`
+   */
+  getArgument: <A extends FunctionArgName<Fn>>(name: A) => ExpressionAstArgument[] | undefined;
+  /**
+   * Overwrites an existing argument with a new value.
+   * In order to support multi-args, the value given must always be
+   * an array.
+   *
+   * @param name The name of the argument to replace.
+   * @param value The value of the argument. Must always be an array.
+   * @return `this`
+   */
+  replaceArgument: <A extends FunctionArgName<Fn>>(
+    name: A,
+    value: Array<FunctionArgs<Fn>[A]>
+  ) => this;
+  /**
+   * Removes an (optional) argument from the function.
+   *
+   * TypeScript will enforce that you only remove optional
+   * arguments. For manipulating required args, use `replaceArgument`.
+   *
+   * @param name The name of the argument to remove.
+   * @return `this`
+   */
+  removeArgument: <A extends OptionalKeys<FunctionArgs<Fn>>>(name: A) => this;
+  /**
+   * Converts function to an AST.
+   *
+   * @return `ExpressionAstFunction`
+   */
+  toAst: () => ExpressionAstFunction;
+}
+
+/**
+ * Manages an AST for a single expression function. The return value
+ * can be provided to `buildExpression` to add this function to an
+ * expression.
+ *
+ * Note that to preserve type safety and ensure no args are missing,
+ * all required arguments for the specified function must be provided
+ * up front. If desired, they can be changed or removed later.
+ *
+ * @param fnName String representing the name of this expression function.
+ * @param initialArgs Object containing the arguments to this function.
+ * @return `this`
+ */
+export function buildExpressionFunction<F extends string>(
+  fnName: F,
+  initialArgs: FunctionArgs<F>
+): ExpressionAstFunctionBuilder<F> {
+  const args: ExpressionAstFunction['arguments'] = initialArgs;
+  Object.entries(initialArgs).forEach(([key, value]) => {
+    args[key] = Array.isArray(value) ? value : [value];
+  });
+
+  return {
+    name: fnName,
+    arguments: args,
+
+    addArgument(key, value) {
+      if (args.hasOwnProperty(key)) {
+        throw new Error('Argument already exists on this function');
+      }
+      args[key] = Array.isArray(value) ? value : [value];
+      return this;
+    },
+
+    getArgument(key) {
+      if (!args.hasOwnProperty(key)) {
+        return;
+      }
+      return args[key];
+    },
+
+    replaceArgument(key, value) {
+      if (!args.hasOwnProperty(key)) {
```

Comment:
Yeah I had originally called this `updateArgument`, and then renamed to `replaceArgument` because I thought it was clearer. For example,

```ts
fn.addArgument('foo', [0,1]);
fn.updateArgument('foo', 2);
```

In the above case it feels unclear whether `2` is getting pushed into the array of args, or is completely replacing the existing values.

So in the current implementation, to manipulate single values in a multi arg, you would need to retrieve the array of that arg's values, and manipulate it directly. So to add an arg:

```ts
      const fn = buildExpressionFunction('hello', { world: [0, 1] });
      const arg = fn.getArgument('world');
      if (arg) {
        arg.push(2);
      }
      fn.getArgument('world'); // [0, 1, 2]
```

Or to update:

```ts
      const arg = fn.getArgument('world'); // [0, 1, 2]
      if (arg) {
        fn.replaceArgument(
          'world',
          arg.filter(a => a !== 1)
        );
      }
```

Reply:
I wasn't sure how often we would have this use case, but I could see eventually adding convenience methods for some of the above functionality if we think they will be common use cases.

Reply:
I've updated this slightly now, so that `addArgument` will automatically push any new entry to the multi-arg array. 

Manipulating the args as I've suggested still works, however it isn't providing you with any type safety as you are messing with the AST directly. (And it also doesn't handle the auto conversion of subexpressions to AST).

So `addArgument` works now, the only tradeoff is that folks need to take care not to keep calling it multiple times on the same arg unless they are certain it supports `multi: true`. I added a TODO so we can think about whether there are ways to better type check this case.